usr.bin/lock/lock.1
.\"	$NetBSD$
.\"
.\" Copyright (c) 1987, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)lock.1	8.1 (Berkeley) 6/6/93
.\"
.Dd June 6, 1993
.Dt LOCK 1
.Os
.Sh NAME
.Nm lock
.Nd reserve a terminal
.Sh SYNOPSIS
.Nm
.Op Fl np
.Op Fl t Ar timeout
.Sh DESCRIPTION
.Nm
requests a password from the user, reads it again for verification
and then will normally not relinquish the terminal until the password is
repeated.
There are two other conditions under which it will terminate: it
will timeout after some interval of time and it may be killed by someone
with the appropriate permission.
.Pp
Options:
.Pp
.Bl -tag -width Fl
.It Fl n
No timeout is used.
The terminal will be locked indefinitely or until current challenge is met.
.It Fl p
A password is not requested, instead the user's current login password
is used.
If the user has an S/Key key, they may also use it
to unlock the terminal.
To do this the user should enter "s/key" at the unlock "Key:" prompt.
The user will then be issued an S/Key challenge to which they may respond
with a six-word S/Key one-time password.
.It Fl t Ar timeout
The time limit (default 15 minutes) is changed to
.Ar timeout
minutes.
.El
.Sh SEE ALSO
.Xr skey 1
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

usr.bin/lock/lock.c
/*	$NetBSD$	*/

/*
 * Copyright (c) 1980, 1987, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Bob Toxen.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1987, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)lock.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD$");
#endif /* not lint */

/*
 * Lock a terminal up until the given key is entered, until the root
 * password is entered, or the given interval times out.
 *
 * Timeout interval is by default TIMEOUT, it can be changed with
 * an argument of the form -time where time is in minutes
 */

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <signal.h>

#include <err.h>
#include <pwd.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#ifdef USE_PAM
#include <security/pam_appl.h>
#include <security/openpam.h>	/* for openpam_ttyconv() */
#endif

#ifdef SKEY
#include <skey.h>
#endif


#define	TIMEOUT	15

int	main(int, char **);

static void	bye(int) __dead;
static void	hi(int);
static void	quit(int) __dead;
#ifdef SKEY
static int	skey_auth(const char *);
#endif

static struct timeval	timeout;
static struct timeval	zerotime;
static struct termios	tty, ntty;
static int	notimeout;			/* no timeout at all */
static long	nexttime;			/* keep the timeout time */

int
main(int argc, char **argv)
{
	struct passwd *pw;
	struct timeval timval;
	struct itimerval ntimer, otimer;
	struct tm *timp;
	time_t curtime;
	int ch, usemine;
	long sectimeout;
	char *ap, *ttynam;
	const char *tzn;
	uid_t uid = getuid();
	char hostname[MAXHOSTNAMELEN + 1], s[BUFSIZ], s1[BUFSIZ];
#ifdef USE_PAM
	pam_handle_t *pamh = NULL;
	static const struct pam_conv pamc = { &openpam_ttyconv, NULL };
	int pam_err;
#else
	char *mypw = NULL;
#endif

	if ((pw = getpwuid(getuid())) == NULL)
		errx(1, "unknown uid %lu.", (u_long)uid);

	notimeout = 0;
	sectimeout = TIMEOUT;
	usemine = 0;

	while ((ch = getopt(argc, argv, "npt:")) != -1)
		switch ((char)ch) {
		case 'n':
			notimeout = 1;
			break;
		case 't':
			errno = 0;
			if (((sectimeout = strtol(optarg, &ap, 0)) == LONG_MAX
			    || sectimeout == LONG_MIN)
			    && errno == ERANGE)
				err(1, "illegal timeout value: %s", optarg);
			if (optarg == ap || *ap || sectimeout <= 0)
				errx(1, "illegal timeout value: %s", optarg);
			if (sectimeout >= INT_MAX / 60)
				errx(1, "too large timeout value: %ld",
				    sectimeout);
			break;
		case 'p':
			usemine = 1;
#ifndef USE_PAM
			mypw = strdup(pw->pw_passwd);
			if (!mypw)
				err(1, "strdup");
#endif
			break;
		case '?':
		default:
			(void)fprintf(stderr,
			    "usage: %s [-np] [-t timeout]\n", getprogname());
			exit(1);
		}

#if defined(USE_PAM) || defined(SKEY)
	if (! usemine) {	/* -p with PAM or S/key needs privs */
#endif
	if (setuid(uid) == -1)	/* discard privs */
		err(1, "setuid failed");
#if defined(USE_PAM) || defined(SKEY)
	}
#endif

	timeout.tv_sec = (int)sectimeout * 60;

	if (tcgetattr(STDIN_FILENO, &tty) < 0)	/* get information for header */
		err(1, "tcgetattr failed");
	gethostname(hostname, sizeof(hostname));
	hostname[sizeof(hostname) - 1] = '\0';
	if (!(ttynam = ttyname(STDIN_FILENO)))
		err(1, "ttyname failed");
	if (gettimeofday(&timval, NULL) == -1)
		err(1, "gettimeofday failed");
	curtime = timval.tv_sec;
	nexttime = timval.tv_sec + ((int)sectimeout * 60);
	timp = localtime(&curtime);
	ap = asctime(timp);
#ifdef __SVR4
	tzn = tzname[0];
#else
	tzn = timp->tm_zone;
#endif

	if (signal(SIGINT, quit) == SIG_ERR)
		err(1, "signal failed");
	if (signal(SIGQUIT, quit) == SIG_ERR)
		err(1, "signal failed");
	ntty = tty; ntty.c_lflag &= ~ECHO;
	if (tcsetattr(STDIN_FILENO, TCSADRAIN, &ntty) == -1)
		err(1, "tcsetattr");

	if (!usemine) {
		/* get key and check again */
		(void)printf("Key: ");
		if (!fgets(s, sizeof(s), stdin) || *s == '\n')
			quit(0);
		(void)printf("\nAgain: ");
		/*
		 * Don't need EOF test here, if we get EOF, then s1 != s
		 * and the right things will happen.
		 */
		(void)fgets(s1, sizeof(s1), stdin);
		(void)putchar('\n');
		if (strcmp(s1, s)) {
			(void)printf("\alock: passwords didn't match.\n");
			(void)tcsetattr(STDIN_FILENO, TCSADRAIN, &tty);
			exit(1);
		}
		s[0] = '\0';
#ifndef USE_PAM
		mypw = s1;
#endif
	}
#ifdef USE_PAM
	if (usemine) {
		pam_err = pam_start("lock", pw->pw_name, &pamc, &pamh);
		if (pam_err != PAM_SUCCESS)
			err(1, "pam_start: %s", pam_strerror(NULL, pam_err));
	}
#endif

	/* set signal handlers */
	if (signal(SIGINT, hi) == SIG_ERR)
		err(1, "signal failed");
	if (signal(SIGQUIT, hi) == SIG_ERR)
		err(1, "signal failed");
	if (signal(SIGTSTP, hi) == SIG_ERR)
		err(1, "signal failed");

	if (notimeout) {
		if (signal(SIGALRM, hi) == SIG_ERR)
			err(1, "signal failed");
		(void)printf("lock: %s on %s.  no timeout.\n"
		    "time now is %.20s%s%s",
		    ttynam, hostname, ap, tzn, ap + 19);
	}
	else {
		if (signal(SIGALRM, bye) == SIG_ERR)
			err(1, "signal failed");

		ntimer.it_interval = zerotime;
		ntimer.it_value = timeout;
		if (setitimer(ITIMER_REAL, &ntimer, &otimer) == -1)
			err(1, "setitimer failed");

		/* header info */
		(void)printf("lock: %s on %s. timeout in %ld minutes\n"
		    "time now is %.20s%s%s",
		    ttynam, hostname, sectimeout, ap, tzn, ap + 19);
	}

	for (;;) {
#ifdef USE_PAM
		if (usemine) {
			pam_err = pam_authenticate(pamh, 0);
			if (pam_err == PAM_SUCCESS)
				break;
			goto tryagain;
		}
#endif
		(void)printf("Key: ");
		if (!fgets(s, sizeof(s), stdin)) {
			clearerr(stdin);
			hi(0);
			goto tryagain;
		}
#ifndef USE_PAM
		if (usemine) {
			s[strlen(s) - 1] = '\0';
#ifdef SKEY
			if (strcasecmp(s, "s/key") == 0) {
				if (skey_auth(pw->pw_name))
					break;
			}
#endif
			if (!strcmp(mypw, crypt(s, mypw)))
				break;
		}
		else
#endif
			if (!strcmp(s, s1))
				break;
		(void)printf("\a\n");
 tryagain:
		if (tcsetattr(STDIN_FILENO, TCSADRAIN, &ntty) == -1
		    && errno != EINTR)
			err(1, "tcsetattr failed");
	}
#ifdef USE_PAM
	if (usemine) {
		(void)pam_end(pamh, pam_err);
	}
#endif
	quit(0);
	/* NOTREACHED */
	return 0;
}

#ifdef SKEY
/*
 * We can't use libskey's skey_authenticate() since it
 * handles signals in a way that's inappropriate
 * for our needs. Instead we roll our own.
 */
static int
skey_auth(const char *user)
{
	char s[128];
	const char *ask;
	int ret = 0;

	if (!skey_haskey(user) && (ask = skey_keyinfo(user))) {
		(void)printf("\n[%s]\nResponse: ", ask);
		if (!fgets(s, sizeof(s), stdin) || *s == '\n')
			clearerr(stdin);
		else {
			s[strlen(s) - 1] = '\0';
			if (skey_passcheck(user, s) != -1)
				ret = 1;
		}
	} else
		(void)printf("Sorry, you have no s/key.\n");
	return ret;
}
#endif

static void
hi(int dummy)
{
	struct timeval timval;

	if (notimeout)
		(void)printf("lock: type in the unlock key.\n");
	else {
		if (gettimeofday(&timval, NULL) == -1)
			err(1, "gettimeofday failed");
		(void)printf("lock: type in the unlock key. "
		    "timeout in %lld:%lld minutes\n",
		    (long long)(nexttime - timval.tv_sec) / 60,
		    (long long)(nexttime - timval.tv_sec) % 60);
	}
}

static void
quit(int dummy)
{
	(void)putchar('\n');
	(void)tcsetattr(STDIN_FILENO, TCSADRAIN, &tty);
	exit(0);
}

static void
bye(int dummy)
{
	(void)tcsetattr(STDIN_FILENO, TCSADRAIN, &tty);
	(void)printf("lock: timeout\n");
	exit(1);
}
